File: aoc2024/aoc8.hs
import Control.Monad (guard)
import Data.HashMap.Strict qualified as HM
import Data.HashSet qualified as HS
import Data.List (tails)
import System.Environment (getArgs)

repeatWhile :: (a -> a) -> (a -> Bool) -> a -> [a]
repeatWhile _ condition start | not (condition start) = []
repeatWhile mkNext condition start = start : repeatWhile mkNext condition (mkNext start)

both2 :: (a -> b -> c) -> (a, a) -> (b, b) -> (c, c)
both2 fn (x1, x2) (y1, y2) = (fn x1 y1, fn x2 y2)

main :: IO ()
main = do
  args <- getArgs
  let filename = if null args then "aoc8.in" else head args
  grid <- lines <$> readFile filename
  let antennas = HM.fromListWith (++) $ do
        (idx, row) <- zip [(0 :: Int) ..] grid
        (cidx, spot) <- zip [(0 :: Int) ..] row
        guard (spot /= '.')
        pure (spot, [(idx, cidx)])
  let height = length grid
  let width = length $ head grid
  let isInside (x, y) = (0 <= x) && (0 <= y) && (x < height) && (y < width)
  let antinodes1 = HS.toList $ HS.fromList $ filter isInside $ do
        (_, antennas') <- HM.toList antennas
        (spot1 : rest) <- tails antennas'
        spot2 <- rest
        let twoAMinB a b = 2 * a - b
        [both2 twoAMinB spot1 spot2, both2 twoAMinB spot2 spot1]
  putStrLn $ "Part 1: " ++ show (length antinodes1)
  let antinodes2 = HS.toList $ HS.fromList $ filter isInside $ do
        (_, antennas') <- HM.toList antennas
        (spot1 : rest) <- tails antennas'
        spot2 <- rest
        let nxt1 = both2 (+) $ both2 (-) spot1 spot2
        let nxt2 = both2 (+) $ both2 (-) spot2 spot1
        let left = repeatWhile nxt1 isInside spot1
        let right = repeatWhile nxt2 isInside spot2
        left ++ right
  putStrLn $ "Part 2: " ++ show (length antinodes2)

File: aoc2024/aoc8.py
from aoc_util import get_data_lines
import itertools

data = get_data_lines(8)

antennas: dict[str, list[tuple[int, int]]] = {}
width = len(data[0])
height = len(data)
for idx, line in enumerate(data):
    for cidx, char in enumerate(line):
        if char != ".":
            antennas.setdefault(char, []).append((idx, cidx))

antinodes = set()
for color, locs in antennas.items():
    for spot1, spot2 in itertools.combinations(locs, 2):
        antinode1 = (2 * spot1[0] - spot2[0], 2 * spot1[1] - spot2[1])
        antinode2 = (2 * spot2[0] - spot1[0], 2 * spot2[1] - spot1[1])
        if (0 <= antinode1[0] < height) and (0 <= antinode1[1] < width):
            antinodes.add(antinode1)
        if (0 <= antinode2[0] < height) and (0 <= antinode2[1] < width):
            antinodes.add(antinode2)

print("Part 1:", len(antinodes))

antinodes = set()
for color, locs in antennas.items():
    for spot1, spot2 in itertools.combinations(locs, 2):
        anode = spot1
        while (0 <= anode[0] < height) and (0 <= anode[1] < width):
            antinodes.add(anode)
            anode = (anode[0] + spot1[0] - spot2[0], anode[1] + spot1[1] - spot2[1])
        anode = spot2
        while (0 <= anode[0] < height) and (0 <= anode[1] < width):
            antinodes.add(anode)
            anode = (anode[0] + spot2[0] - spot1[0], anode[1] + spot2[1] - spot1[1])

print("Part 2:", len(antinodes))
